
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>API Documentation &#8212; INQuant 2.0 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=8f2a1f02" />
  
  <!-- So that users can add custom icons -->
  <script src="_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="_static/documentation_options.js?v=60dbed4a"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'modules';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Example Usage" href="examples.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
  
    <p class="title logo__title">INQuant 2.0 documentation</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="user_guide.html">
    User Guide
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="#">
    API Documentation
  </a>
</li>

  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="user_guide.html">
    User Guide
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="#">
    API Documentation
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"></div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">API Documentation</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="api-documentation">
<h1>API Documentation<a class="headerlink" href="#api-documentation" title="Link to this heading">#</a></h1>
<section id="module-inquant_tools.Calibration">
<span id="calibration"></span><h2>Calibration<a class="headerlink" href="#module-inquant_tools.Calibration" title="Link to this heading">#</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="inquant_tools.Calibration.Calibration">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">inquant_tools.Calibration.</span></span><span class="sig-name descname"><span class="pre">Calibration</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">predictions_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#inquant_tools.Calibration.Calibration" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="inquant_tools.Calibration.Calibration.load_mzml">
<span class="sig-name descname"><span class="pre">load_mzml</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mzml_path=&lt;class</span> <span class="pre">'str'&gt;</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#inquant_tools.Calibration.Calibration.load_mzml" title="Link to this definition">#</a></dt>
<dd><p>Loads the mzML file and extracts spectra data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mzML_path</strong> (<em>str</em>) – The path to the mzML file to be loaded.</p>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="inquant_tools.Calibration.Calibration.exp">
<span class="sig-name descname"><span class="pre">exp</span></span><a class="headerlink" href="#inquant_tools.Calibration.Calibration.exp" title="Link to this definition">#</a></dt>
<dd><p>An object to store the loaded mzML data.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>MSExperiment</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="inquant_tools.Calibration.Calibration.spectra_data">
<span class="sig-name descname"><span class="pre">spectra_data</span></span><a class="headerlink" href="#inquant_tools.Calibration.Calibration.spectra_data" title="Link to this definition">#</a></dt>
<dd><p>A dictionary to store spectra data, including measured m/z, retention time, peaks, and charge.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="inquant_tools.Calibration.Calibration.weight">
<span class="sig-name descname"><span class="pre">weight</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">grid_point</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_adjustment=&lt;class</span> <span class="pre">'bool'&gt;</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#inquant_tools.Calibration.Calibration.weight" title="Link to this definition">#</a></dt>
<dd><p>Calculates the weights for the grid points based on the given grid point.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>grid_point</strong> (<em>list</em>) – A list containing the grid point coordinates for a data point.</p></li>
<li><p><strong>return_adjustment</strong> (<em>bool</em>) – If True, returns the adjustment factor.
If False, returns the weightsn for each grid point.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>If return_adjustment is True, returns the adjustment factor.
If return_adjustment is False, returns a list of weights for each grid point.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list or float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="inquant_tools.Calibration.Calibration.calibration">
<span class="sig-name descname"><span class="pre">calibration</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mzML_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rt_grids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mz_grids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">set_training_seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#inquant_tools.Calibration.Calibration.calibration" title="Link to this definition">#</a></dt>
<dd><p>Calibrates the predictions based on the mzML file and the specified parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mzML_id</strong> (<em>str</em>) – The ID of the mzML file to be calibrated.</p></li>
<li><p><strong>train_size</strong> (<em>float</em><em>, </em><em>optional</em>) – The fraction of data to be used for training (default is 0.10).</p></li>
<li><p><strong>rt_grids</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of bins for the retention time grid (default is 50).</p></li>
<li><p><strong>mz_grids</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of bins for the mass-to-charge ratio grid (default is 200).</p></li>
<li><p><strong>set_training_seed</strong> (<em>int</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – The seed for random number generation (default is None).</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="inquant_tools.Calibration.Calibration.dev_matrix">
<span class="sig-name descname"><span class="pre">dev_matrix</span></span><a class="headerlink" href="#inquant_tools.Calibration.Calibration.dev_matrix" title="Link to this definition">#</a></dt>
<dd><p>A dictionary to store the deviation matrix for calibration.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="inquant_tools.Calibration.Calibration.total_weight">
<span class="sig-name descname"><span class="pre">total_weight</span></span><a class="headerlink" href="#inquant_tools.Calibration.Calibration.total_weight" title="Link to this definition">#</a></dt>
<dd><p>A dictionary to store the total possible weight for each grid point.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="inquant_tools.Calibration.Calibration.grid_zero">
<span class="sig-name descname"><span class="pre">grid_zero</span></span><a class="headerlink" href="#inquant_tools.Calibration.Calibration.grid_zero" title="Link to this definition">#</a></dt>
<dd><p>An array to store the zero point of the grid.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="inquant_tools.Calibration.Calibration.grid_basis">
<span class="sig-name descname"><span class="pre">grid_basis</span></span><a class="headerlink" href="#inquant_tools.Calibration.Calibration.grid_basis" title="Link to this definition">#</a></dt>
<dd><p>An array to store the basis of the grid.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="inquant_tools.Calibration.Calibration.get_ppm_tolerance">
<span class="sig-name descname"><span class="pre">get_ppm_tolerance</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#inquant_tools.Calibration.Calibration.get_ppm_tolerance" title="Link to this definition">#</a></dt>
<dd><p>Returns the maximum ppm tolerance from the list of ppm tolerances. This is the recommended m/z tolerance for calculating quantification in INQuant.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>max_ppm_tolerance</strong> – The maximum ppm tolerance from all the predictions.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="inquant_tools.Calibration.Calibration.write_calibrated_mzml_file">
<span class="sig-name descname"><span class="pre">write_calibrated_mzml_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mzml_name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#inquant_tools.Calibration.Calibration.write_calibrated_mzml_file" title="Link to this definition">#</a></dt>
<dd><p>Writes the calibrated predictions to a CSV file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_path</strong> (<em>str</em>) – The path where the calibrated mzML files will be saved.</p></li>
<li><p><strong>mzml_name</strong> (<em>str</em>) – The ID of the mzML file to be saved.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>new_predictions</strong> (<em>str</em>) – The path of the new calibrated predictions CSV file.</p></li>
<li><p><strong>new_mzml</strong> (<em>str</em>) – The path of the new mzML file with calibrated spectra.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="inquant_tools.Calibration.Calibration.write_calibrated_predictions_file">
<span class="sig-name descname"><span class="pre">write_calibrated_predictions_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">predictions_name=&lt;class</span> <span class="pre">'str'&gt;</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#inquant_tools.Calibration.Calibration.write_calibrated_predictions_file" title="Link to this definition">#</a></dt>
<dd><p>Writes the calibrated predictions to a CSV file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_path</strong> (<em>str</em>) – The path where the calibrated predictions CSV file will be saved.</p></li>
<li><p><strong>predictions_name</strong> (<em>str</em>) – The name of the predictions file to be saved.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>new_predictions</strong> – The path of the new calibrated predictions CSV file.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-inquant_tools.INQuant">
<span id="inquant"></span><h2>INQuant<a class="headerlink" href="#module-inquant_tools.INQuant" title="Link to this heading">#</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="inquant_tools.INQuant.INQuant">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">inquant_tools.INQuant.</span></span><span class="sig-name descname"><span class="pre">INQuant</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">predictions_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mzml_file_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proteome_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">script_purpose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ppm_tolerance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">empty_values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">confidence_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.95</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">experiment_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intensity_tolerance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rt_tolerance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.01</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#inquant_tools.INQuant.INQuant" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="inquant_tools.INQuant.INQuant.INQuant_close">
<span class="sig-name descname"><span class="pre">INQuant_close</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#inquant_tools.INQuant.INQuant.INQuant_close" title="Link to this definition">#</a></dt>
<dd><p>Exit function writes the status file for the experiment, should be run if the script is run without using the with-statement.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="inquant_tools.INQuant.INQuant.time_flag">
<span class="sig-name descname"><span class="pre">time_flag</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#inquant_tools.INQuant.INQuant.time_flag" title="Link to this definition">#</a></dt>
<dd><p>Records the time taken since the start of the script and appends the
result to the timer counts. If a <cite>time_flag</cite> is set anywhere within
the with-statement, the status file will show how long the script took
to reach the <cite>time_flag</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>name</strong> (<em>str</em>) – The name associated with the time flag. This will be used in the status
file to indicate the specific point in the script.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>This method does not return any value. It appends the time information
to the <cite>timer_counts</cite> list attribute.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="inquant_tools.INQuant.INQuant.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">noise_boundary</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">noise_ms_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mbr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mbr_tolerance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.9</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cleavage_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quant_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_n_peptides</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_abundance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'median'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write_psm_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">psm_file_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write_peptide_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">peptide_file_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write_protein_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protein_file_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write_ungrouped_protein_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ungrouped_protein_file_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write_individual_quantification_files</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individual_quant_file_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite_all</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#inquant_tools.INQuant.INQuant.run" title="Link to this definition">#</a></dt>
<dd><p>Perform the full quantification process by combining multiple functionalities
of the class into a single command. This should be executed when the class is
initialized with a <cite>with</cite> statement to ensure proper management of resources.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>noise_boundary</strong> (<em>float</em><em>, </em><em>optional</em>) – The boundary for noise calculation. Default is 0.05.</p></li>
<li><p><strong>noise_ms_level</strong> (<em>int</em><em>, </em><em>optional</em>) – The MS level for noise calculation. Default is 2.</p></li>
<li><p><strong>mbr</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to run the MBR (Match Between Runs) algorithm. Default is False.</p></li>
<li><p><strong>mbr_tolerance</strong> (<em>float</em>) – The minimum confidence for the MBR spectras. Only predictions with a higher confidence
will have mbr matches. Default is 0.9.</p></li>
<li><p><strong>cleavage_length</strong> (<em>int</em><em>, </em><em>default=4</em>) – The length of the cleavage site used for alignment. This parameter is used to output the protein position,
with cleavage length being the number of amino acids before and after the peptide sequence in the protein.</p></li>
<li><p><strong>quant_method</strong> (<em>str</em><em>, </em><em>optional</em>) – The quantification method to use. Default is ‘mean’.</p></li>
<li><p><strong>top_n_peptides</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of top peptides to consider for quantification values for each protein. Default is 5.</p></li>
<li><p><strong>normalize_abundance</strong> (<em>str</em><em>, </em><em>optional</em>) – The method for normalizing abundance values. Options include ‘median’,
‘mean’, ‘tic’, or ‘false’. Default is ‘median’.</p></li>
<li><p><strong>write_protein_table</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to write the protein table to a file. Default is True.</p></li>
<li><p><strong>protein_file_name</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the protein table file. Default is ‘[experiment_name]_protein_table.csv’.</p></li>
<li><p><strong>write_peptide_table</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to write the peptide table to a file. Default is True.</p></li>
<li><p><strong>peptide_file_name</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the peptide table file. Default is ‘[experiment_name]_peptide_table.csv’.</p></li>
<li><p><strong>write_psm_table</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to write the PSM (Peptide-Spectrum Match) table to a file. Default is True.</p></li>
<li><p><strong>psm_file_name</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the PSM table file. Default is ‘[experiment_name]_psm_table.csv’.</p></li>
<li><p><strong>write_ungrouped_protein_table</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to write the ungrouped protein table to a file. Default is False.</p></li>
<li><p><strong>ungrouped_protein_file_name</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the ungrouped protein table file. Default is ‘[experiment_name]_ungrouped_protein_table.csv’.</p></li>
<li><p><strong>write_individual_quantification_files</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to write individual quantification files for each mzML file. Default is False.</p></li>
<li><p><strong>individual_quant_file_name</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the individual quantification file. Default is ‘[experiment_name]_quantification_[mzml_file_id].csv’.</p></li>
<li><p><strong>overwrite_all</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to overwrite all existing files without prompting. Default is False,
which will ask for confirmation before overwriting any file.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the <cite>top_n_peptides</cite> parameter is less than 1, a <cite>ValueError</cite> will be raised.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="inquant_tools.INQuant.INQuant.load_predictions">
<span class="sig-name descname"><span class="pre">load_predictions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mbr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mbr_tolerance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.9</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="headerlink" href="#inquant_tools.INQuant.INQuant.load_predictions" title="Link to this definition">#</a></dt>
<dd><p>Loads predictions into the class and assigns file IDs for all mzml files in the experiment.
If only one mzml file is provided, the file ID will always be ‘mzml’. This function handles
multiple mzml files and attempts to match their identifiers to those in the predictions file.
Additionally, if mbr (Match Between Runs) is enabled, it processes missing data by
imputing values from previous replicates.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mbr</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default is False</em>) – Flag indicating whether to run mbr (Match Between Runs). If True, missing values
for certain replicates will be imputed based on available data from other replicates.</p></li>
<li><p><strong>mbr_tolerance</strong> (<em>float</em><em>, </em><em>default is 0.9</em>) – The minimum confidence for the mbr specs. Only predictions with a higher confidence
will have mbr matches.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>sorted_predictions_dict</strong> – The function modifies internal class attributes such as <cite>mzml_file_dict</cite>, <cite>sorted_predictions_df</cite>,
and <cite>sorted_predictions_dict</cite>. If <cite>mbr</cite> is True, the method also populates the mbr specifications
into <cite>sorted_predictions_dict</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>Break :</strong> – If there is an error matching mzml files with prediction IDs or other issues arise, the process
    will stop and the Break exception will be raised.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="inquant_tools.INQuant.INQuant.make_psm_table">
<span class="sig-name descname"><span class="pre">make_psm_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">predictions_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">noise_boundary</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">noise_ms_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="headerlink" href="#inquant_tools.INQuant.INQuant.make_psm_table" title="Link to this definition">#</a></dt>
<dd><p>Creates a Peptide-Spectrum Match (PSM) table from the predictions data.
This function is dependent on the <cite>load_predictions</cite> method and requires the
predictions data to be loaded first, because of the mzML IDs which are generated
in the <cite>load_predictions</cite> method.
Quantifications use user specifications from the initialization of the class.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>predictions_dict</strong> (<em>dict</em>) – Dictionary containing the predictions data. The keys are mzml file IDs and the
values are DataFrames containing the predictions for each file.</p></li>
<li><p><strong>noise_boundary</strong> (<em>float</em><em>, </em><em>optional</em><em>, </em><em>default is 0.05</em>) – Boundary value for noise calculation. This value will be used to define the
threshold below which signals are considered as noise.</p></li>
<li><p><strong>noise_ms_level</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default is 2</em>) – MS level to be used for noise calculation. Determines which level of the
mass spectrometry data (MS1 or MS2) will be considered for noise analysis.</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="inquant_tools.INQuant.INQuant.noise_variables">
<span class="sig-name descname"><span class="pre">noise_variables</span></span><a class="headerlink" href="#inquant_tools.INQuant.INQuant.noise_variables" title="Link to this definition">#</a></dt>
<dd><p>String containing the noise calculation parameters used in the experiment.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>psm_dict</strong> – Dictionary containing the PSM data with the calculated quantification values.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>The noise calculation is based on the provided boundary and MS level, which helps in filtering out irrelevant signals.</p></li>
<li><p>Noise for each experiment can be found in the status file, if such a file is generated by using the with-statement for the class.</p></li>
</ul>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="inquant_tools.INQuant.INQuant.make_peptide_table">
<span class="sig-name descname"><span class="pre">make_peptide_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">psm_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="headerlink" href="#inquant_tools.INQuant.INQuant.make_peptide_table" title="Link to this definition">#</a></dt>
<dd><p>Pivots the quantification data into a wide format, where the abundance for each peptide
in the experiment is moved to a column with a corresponding name. The columns are sorted
to combine identical peptides with mass charges within a specified tolerance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>psm_dict</strong> (<em>dict</em>) – A dict containing the quantification data for the peptides. This dict should
include columns such as ‘sequence’, ‘charge’, and ‘abundance’ to be pivoted into the wide format.</p>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="inquant_tools.INQuant.INQuant.peptide_dict">
<span class="sig-name descname"><span class="pre">peptide_dict</span></span><a class="headerlink" href="#inquant_tools.INQuant.INQuant.peptide_dict" title="Link to this definition">#</a></dt>
<dd><p>Dictionary containing peptide information.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>peptide_dict</strong> – A dictionary containing the quantification data for the peptides in a wide format.
The keys are the peptide sequences and the values are dictionaries with the peptide
data and corresponding abundance values for each experiment.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>The function combines peptides with mass charge ratios within tolerance.</p></li>
<li><p>This function also updates the <cite>self.peptide_dict</cite> attribute of the class with the new data.</p></li>
</ul>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="inquant_tools.INQuant.INQuant.normalizer">
<span class="sig-name descname"><span class="pre">normalizer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptide_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'median'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="headerlink" href="#inquant_tools.INQuant.INQuant.normalizer" title="Link to this definition">#</a></dt>
<dd><p>Normalizes the quantification data based on the specified method. The default method is
‘median’, where the column with the highest amount of data points is used as the baseline.
Other available methods are ‘mean’ and ‘tic’ (Total Ion Current).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptide_dict</strong> (<em>dict</em>) – A dictionary containing the quantification data for the peptides in wide format.</p></li>
<li><p><strong>type</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default='median'</em>) – <p>The normalization method to use. Options are:</p>
<ul>
<li><p>’median’ : Normalize using the median of the data.</p></li>
<li><p>’mean’ : Normalize using the mean of the data.</p></li>
<li><p>’tic’ : Normalize using the Total Ion Current method.</p></li>
</ul>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>normalized_dict</strong> – The normalized peptide dict with additional values for the normalized abundance values.
The new values are named with a suffix ‘_normalized’ to indicate the normalization.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>The normalization method adjusts the abundance values to correct for systematic biases.</p></li>
<li><p>The ‘tic’ method sums all intensities in a sample and normalizes each peptide’s abundance by the total sum.</p></li>
<li><p>This method also updates the <cite>self.peptide_dict</cite> attribute with the normalized values.</p></li>
</ul>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="inquant_tools.INQuant.INQuant.load_proteome">
<span class="sig-name descname"><span class="pre">load_proteome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">proteome_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#inquant_tools.INQuant.INQuant.load_proteome" title="Link to this definition">#</a></dt>
<dd><p>Loads a proteome file (in FASTA format) into a dictionary, where the key is a tuple
of the protein’s ID and description, and the value is the protein sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>proteome_file</strong> (<em>str</em>) – The file path to the proteome file in FASTA format. The file should contain protein
sequences, with each entry beginning with a ‘&gt;’ symbol followed by the protein’s
ID and description, and then the sequence itself on the next lines.</p>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="inquant_tools.INQuant.INQuant.proteome_dict">
<span class="sig-name descname"><span class="pre">proteome_dict</span></span><a class="headerlink" href="#inquant_tools.INQuant.INQuant.proteome_dict" title="Link to this definition">#</a></dt>
<dd><p>Dictionary containing information from the proteome file.
The keys are tuples of the protein’s ID and description, and the values are the protein sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="inquant_tools.INQuant.INQuant.proteome_description_dict">
<span class="sig-name descname"><span class="pre">proteome_description_dict</span></span><a class="headerlink" href="#inquant_tools.INQuant.INQuant.proteome_description_dict" title="Link to this definition">#</a></dt>
<dd><p>Dictionary containing the protein descriptions from the proteome file.
The keys are tuples of the protein’s ID and description, and the values are the protein descriptions.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The function updates the class attributes <cite>self.proteome_dict</cite> and
<cite>self.proteome_description_dict</cite> with the loaded protein sequences and their
corresponding descriptions.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>The FASTA format should be well-structured, where each protein entry starts with a ‘&gt;’ symbol, followed by an identifier and description (separated by spaces), and the sequence appears on the following lines.</p></li>
<li><p>If the proteome file contains multiple sequences, each will be parsed and stored in the dictionary.</p></li>
</ul>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="inquant_tools.INQuant.INQuant.compute_alignments">
<span class="sig-name descname"><span class="pre">compute_alignments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptide_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cleavage_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="headerlink" href="#inquant_tools.INQuant.INQuant.compute_alignments" title="Link to this definition">#</a></dt>
<dd><p>Runs alignment on the peptide sequences against the supplied FASTA file proteome.
Updates the peptide table with information about the number of proteins each peptide
aligns to, the corresponding protein accessions, and the protein matches.
Initializes the <cite>self.protein_dict</cite> attribute to store protein information when for the
making of the protein table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptide_dict</strong> (<em>dict</em><em> or </em><em>path to csv/xlsx file</em><em>, </em><em>optional</em>) – A dictionary representing the peptide table. If not provided, the function uses
the class attribute <cite>self.peptide_dict</cite>. Otherwise, it attempts to load the
peptide table from the specified file path.</p></li>
<li><p><strong>cleavage_length</strong> (<em>int</em><em>, </em><em>default=4</em>) – The length of the cleavage site used for alignment. This parameter is used to output the protein position,
with cleavage length being the number of amino acids before and after the peptide sequence in the protein.</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="inquant_tools.INQuant.INQuant.protein_dict">
<span class="sig-name descname"><span class="pre">protein_dict</span></span><a class="headerlink" href="#inquant_tools.INQuant.INQuant.protein_dict" title="Link to this definition">#</a></dt>
<dd><p>Dictionary containing information about the proteins have peptide matches.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><ul class="simple">
<li><p><strong>peptide_dict</strong> (<em>dict</em>) – The updated peptide table with additional information about protein matches,
including the number of protein matches, their accessions, and specific matching proteins.</p></li>
<li><p><strong>protein_dict</strong> (<em>dict</em>) – Dictionary containing information about the proteins that have peptide matches.
The keys are tuples of the protein’s ID and description, and the values are dictionaries with protein data: accession number, description, peptides aligned, alignment coverage and the length of the protein.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>The alignment is performed by comparing the peptide sequences against the protein sequences in the provided proteome (FASTA file).</p></li>
<li><p>For each peptide, the number of protein matches, their accessions, and the specific matching proteins are recorded in the <cite>self.peptide_dict</cite>.</p></li>
<li><p>The FASTA format should contain the protein sequences in a valid format with ‘&gt;’ headers denoting protein IDs/descriptions followed by the sequence.</p></li>
</ul>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="inquant_tools.INQuant.INQuant.make_protein_table">
<span class="sig-name descname"><span class="pre">make_protein_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">psm_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">peptide_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protein_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quant_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_n_peptides</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="headerlink" href="#inquant_tools.INQuant.INQuant.make_protein_table" title="Link to this definition">#</a></dt>
<dd><p>Groups proteins into clusters based on their shared peptides and properties.</p>
<p>This function analyzes the relationship between peptides and proteins, and groups proteins
that share similar peptides. The grouping also takes into account quantification methods
and the top N peptides per protein.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>psm_dict</strong> (<em>dict</em><em>, </em><em>optional</em>) – A dictionary containing the PSM data. If not provided, the function uses
the class attribute <cite>self.psm_dict</cite>.</p></li>
<li><p><strong>peptide_dict</strong> (<em>dict</em><em>, </em><em>optional</em>) – A dictionary containing peptide information. If not provided, the function uses
the class attribute <cite>self.peptide_dict</cite>.</p></li>
<li><p><strong>protein_dict</strong> (<em>dict</em><em>, </em><em>optional</em>) – A dictionary containing protein information. If not provided, the function uses
the class attribute <cite>self.protein_dict</cite>.</p></li>
<li><p><strong>quant_method</strong> (<em>str</em><em>, </em><em>optional</em>) – The protein quantification method to use for the groups. Options include ‘mean’ (default),
‘median’.</p></li>
<li><p><strong>top_n_peptides</strong> (<em>int</em><em>, </em><em>optional</em>) – The number peptides to include in the protein quantification. Default is 5.</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="inquant_tools.INQuant.INQuant.grouped_protein_dict">
<span class="sig-name descname"><span class="pre">grouped_protein_dict</span></span><a class="headerlink" href="#inquant_tools.INQuant.INQuant.grouped_protein_dict" title="Link to this definition">#</a></dt>
<dd><p>A dictionary containing the grouped protein information, including the principal protein,
description, protein group, number of peptides, unique peptides, PSMs, and coverage percentage.
Also updates the class attributes <cite>self.psm_dict</cite>, <cite>self.peptide_dict</cite>, <cite>self.grouped_protein_dict</cite>
and <cite>self.protein_dict</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="inquant_tools.INQuant.INQuant.psm_columns">
<span class="sig-name descname"><span class="pre">psm_columns</span></span><a class="headerlink" href="#inquant_tools.INQuant.INQuant.psm_columns" title="Link to this definition">#</a></dt>
<dd><p>A list of the output columns and oclumn order for the PSM table.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="inquant_tools.INQuant.INQuant.peptide_columns">
<span class="sig-name descname"><span class="pre">peptide_columns</span></span><a class="headerlink" href="#inquant_tools.INQuant.INQuant.peptide_columns" title="Link to this definition">#</a></dt>
<dd><p>A list of the output columns and column order for the peptide table.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="inquant_tools.INQuant.INQuant.protein_columns">
<span class="sig-name descname"><span class="pre">protein_columns</span></span><a class="headerlink" href="#inquant_tools.INQuant.INQuant.protein_columns" title="Link to this definition">#</a></dt>
<dd><p>A list of the output columns and column order for the protein table.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>grouped_protein_dict</strong> – A dictionary containing the grouped protein information, including the principal protein,
description, protein group, number of peptides, unique peptides, PSMs, and coverage percentage.
Also updates the class attributes <cite>self.psm_dict</cite>, <cite>self.peptide_dict</cite>, <cite>self.grouped_protein_dict</cite>
and <cite>self.protein_dict</cite>.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>The protein groups are created based on peptide alignment, with quantification methods applied to summarize protein abundances.</p></li>
</ul>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="inquant_tools.INQuant.INQuant.write_files">
<span class="sig-name descname"><span class="pre">write_files</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">write_protein_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protein_file_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write_peptide_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">peptide_file_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write_psm_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">psm_file_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write_ungrouped_protein_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ungrouped_protein_file_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write_individual_quantification_files</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individual_quant_file_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite_all</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#inquant_tools.INQuant.INQuant.write_files" title="Link to this definition">#</a></dt>
<dd><p>Writes the output files for the experiment. Options include writing individual abundance files, protein table,
peptide table, psm table, and grouped protein table. Requires input for the overwrite statement. If none is given,
it will prompt for it.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>write_protein_table</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to write the protein table to a file. Defaults to True.</p></li>
<li><p><strong>protein_file_name</strong> (<em>str</em><em>, </em><em>optional</em>) – Specific file name for the protein table. Defaults to an empty string, in which case the function uses the
default name for the protein file.</p></li>
<li><p><strong>write_peptide_table</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to write the peptide table to a file. Defaults to True.</p></li>
<li><p><strong>peptide_file_name</strong> (<em>str</em><em>, </em><em>optional</em>) – Specific file name for the peptide table. Defaults to an empty string, in which case the function uses the
default name for the peptide file.</p></li>
<li><p><strong>write_psm_table</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to write the psm (Peptide Spectrum Match) table to a file. Defaults to True.</p></li>
<li><p><strong>psm_file_name</strong> (<em>str</em><em>, </em><em>optional</em>) – Specific file name for the psm table. Defaults to an empty string, in which case the function uses the
default name for the psm file.</p></li>
<li><p><strong>write_ungrouped_protein_table</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to write the ungrouped protein table to a file. Defaults to False.</p></li>
<li><p><strong>ungrouped_protein_file_name</strong> (<em>str</em><em>, </em><em>optional</em>) – Specific file name for the ungrouped protein table. Defaults to an empty string, in which case the function uses the
default name for the ungrouped protein file.</p></li>
<li><p><strong>write_individual_quantification_files</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to write individual quantification files for each peptide or protein. Defaults to False.</p></li>
<li><p><strong>individual_quant_file_name</strong> (<em>str</em><em>, </em><em>optional</em>) – Specific prefix for individual quantification files. Defaults to an empty string, in which case the function uses
the default prefix for the quantification files.</p></li>
<li><p><strong>overwrite_all</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to overwrite existing files. If set to True, all existing files will be overwritten without prompt. Defaults
to False, in which case the function will prompt the user for confirmation before overwriting any files.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>This function does not return any value. It writes the output files for the experiment as specified in the
parameters.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>If the <cite>overwrite_all</cite> parameter is set to True, the function will automatically overwrite any existing files without confirmation.</p></li>
<li><p>The function handles multiple output files, depending on the specific options provided by the user.</p></li>
<li><p>If any file names are not specified, default names will be used.</p></li>
</ul>
</dd></dl>

</dd></dl>

</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="examples.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Example Usage</p>
      </div>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-inquant_tools.Calibration">Calibration</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.Calibration.Calibration"><code class="docutils literal notranslate"><span class="pre">Calibration</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.Calibration.Calibration.load_mzml"><code class="docutils literal notranslate"><span class="pre">Calibration.load_mzml()</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.Calibration.Calibration.exp"><code class="docutils literal notranslate"><span class="pre">Calibration.exp</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.Calibration.Calibration.spectra_data"><code class="docutils literal notranslate"><span class="pre">Calibration.spectra_data</span></code></a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.Calibration.Calibration.weight"><code class="docutils literal notranslate"><span class="pre">Calibration.weight()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.Calibration.Calibration.calibration"><code class="docutils literal notranslate"><span class="pre">Calibration.calibration()</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.Calibration.Calibration.dev_matrix"><code class="docutils literal notranslate"><span class="pre">Calibration.dev_matrix</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.Calibration.Calibration.total_weight"><code class="docutils literal notranslate"><span class="pre">Calibration.total_weight</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.Calibration.Calibration.grid_zero"><code class="docutils literal notranslate"><span class="pre">Calibration.grid_zero</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.Calibration.Calibration.grid_basis"><code class="docutils literal notranslate"><span class="pre">Calibration.grid_basis</span></code></a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.Calibration.Calibration.get_ppm_tolerance"><code class="docutils literal notranslate"><span class="pre">Calibration.get_ppm_tolerance()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.Calibration.Calibration.write_calibrated_mzml_file"><code class="docutils literal notranslate"><span class="pre">Calibration.write_calibrated_mzml_file()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.Calibration.Calibration.write_calibrated_predictions_file"><code class="docutils literal notranslate"><span class="pre">Calibration.write_calibrated_predictions_file()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-inquant_tools.INQuant">INQuant</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.INQuant.INQuant"><code class="docutils literal notranslate"><span class="pre">INQuant</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.INQuant.INQuant.INQuant_close"><code class="docutils literal notranslate"><span class="pre">INQuant.INQuant_close()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.INQuant.INQuant.time_flag"><code class="docutils literal notranslate"><span class="pre">INQuant.time_flag()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.INQuant.INQuant.run"><code class="docutils literal notranslate"><span class="pre">INQuant.run()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.INQuant.INQuant.load_predictions"><code class="docutils literal notranslate"><span class="pre">INQuant.load_predictions()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.INQuant.INQuant.make_psm_table"><code class="docutils literal notranslate"><span class="pre">INQuant.make_psm_table()</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.INQuant.INQuant.noise_variables"><code class="docutils literal notranslate"><span class="pre">INQuant.noise_variables</span></code></a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.INQuant.INQuant.make_peptide_table"><code class="docutils literal notranslate"><span class="pre">INQuant.make_peptide_table()</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.INQuant.INQuant.peptide_dict"><code class="docutils literal notranslate"><span class="pre">INQuant.peptide_dict</span></code></a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.INQuant.INQuant.normalizer"><code class="docutils literal notranslate"><span class="pre">INQuant.normalizer()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.INQuant.INQuant.load_proteome"><code class="docutils literal notranslate"><span class="pre">INQuant.load_proteome()</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.INQuant.INQuant.proteome_dict"><code class="docutils literal notranslate"><span class="pre">INQuant.proteome_dict</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.INQuant.INQuant.proteome_description_dict"><code class="docutils literal notranslate"><span class="pre">INQuant.proteome_description_dict</span></code></a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.INQuant.INQuant.compute_alignments"><code class="docutils literal notranslate"><span class="pre">INQuant.compute_alignments()</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.INQuant.INQuant.protein_dict"><code class="docutils literal notranslate"><span class="pre">INQuant.protein_dict</span></code></a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.INQuant.INQuant.make_protein_table"><code class="docutils literal notranslate"><span class="pre">INQuant.make_protein_table()</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.INQuant.INQuant.grouped_protein_dict"><code class="docutils literal notranslate"><span class="pre">INQuant.grouped_protein_dict</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.INQuant.INQuant.psm_columns"><code class="docutils literal notranslate"><span class="pre">INQuant.psm_columns</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.INQuant.INQuant.peptide_columns"><code class="docutils literal notranslate"><span class="pre">INQuant.peptide_columns</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.INQuant.INQuant.protein_columns"><code class="docutils literal notranslate"><span class="pre">INQuant.protein_columns</span></code></a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#inquant_tools.INQuant.INQuant.write_files"><code class="docutils literal notranslate"><span class="pre">INQuant.write_files()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
  </nav></div>

  <div class="sidebar-secondary-item">
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/modules.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2025, Annekatrine Kirketerp-Møller, Ida Sofie Goldschmidt.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.1.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>